File: PA27.java
Leetcode: 54. Spiral Matrix
//Time Complexity : O(n) //n is the toatal no of elements in matrix
//Space Complexity: O(1) // as we are returning that list in result 

class Solution {
    public List<Integer> spiralOrder(int[][] matrix) {
        int l=0;
        if(matrix==null || matrix.length ==0){
            return new ArrayList<>();
        }      
        int r= matrix[0].length -1;
        int t=0;
        int b=matrix.length-1;
        int i=0;
            List<Integer> res= new ArrayList<>();
            while(l<=r && t<=b){
                //res.add(matrix[l])
                
                for(int k=l; k<=r;k++){
                    res.add(matrix[t][k]);
                }
                //r--;
                t++;
                for(int k=t; k<=b;k++){
                    res.add(matrix[k][r]);
                }
                r--;
                if(t<=b){
                    for(int k=r; k>=l;k--){
                        res.add(matrix[b][k]);
                    }
                    b--;
                }
                
                if(r>= l){
                    for(int k=b; k>=t;k--){
                        res.add(matrix[k][l]);
                    }
                    l++;
                }
                
                
                
                
                
            }
        return res;
    }
}

File: PA_27.py
#LC 54 - Spiral Matrix
#Time Complexity - O(row*col)
#Space Complexity - O(row*col)
class Solution(object):
    def spiralOrder(self, matrix):
        """
        :type matrix: List[List[int]]
        :rtype: List[int]
        """
        if not matrix:
            return []
        ans = []
        T,L = 0,0
        B,R = len(matrix)-1,len(matrix[0])-1
        d = 0
        while T <= B and L <= R:
            #Go right
            if d == 0:
                for i in range(L,R+1):
                    ans.append(matrix[T][i])
                T = T + 1
                d = 1
            #Go down
            elif d == 1:
                for i in range(T,B+1):
                    ans.append(matrix[i][R])
                R = R - 1
                d = 2
            #Go Left
            elif d == 2:
                i = R
                while i >= L:
                    ans.append(matrix[B][i])
                    i = i - 1
                B = B - 1
                d = 3
            #Go up
            elif d == 3:
                i = B
                while i >= T:
                    ans.append(matrix[i][L])
                    i = i - 1
                L = L + 1
                d = 0
        return(ans)